mathics/core/expression.py

```diff
@@ -88,21 +88,17 @@
 )
 
 
-def expression_sameQ(self, other):
+def eval_SameQ(self, other):
     """
-    Iterative implementation of SameQ.
+    Iterative implementation of SameQ[].
 
-    Run a tree transversal comparison between `self` and `other`.
+    Tree traversal comparison between `self` and `other`.
     Return `True` if both tree structures are equal.
 
-    This implementation avoids the issue with
-    the recursion limit in Python 3.12
+    This non-recursive implementation reduces the Python stack needed
+    in evaluation. Staring in Python 3.12 there is a limit on the
+    recursion level.
     """
-    # TODO: Consider a faster implementation.
```

Comment:
It is not clear to me this is not fast enough. What is done here is equivalent to tail recursion elimination. 

Actually, on my home machine I've never ran into the recursion limit problem. I guess I have more memory. The reason Python 3.12 put compiler/OS limitation to recursion was no doubt some sort of security thing. People aren't going to know (or care about) what is being referred to with  "the issue".

Reply:
Actually, comparing the time the tests takes in GithubActions, it seems that this is faster than the recursive implementation.